Corona-Warn-App/src/main/java/de/rki/coronawarnapp/task/TaskController.kt

```diff
@@ -0,0 +1,215 @@
+        .onEach { initTaskData(it) }
+        .onCompletion { Timber.tag(TAG).w("Stopped listening to request queue. Why?") }
+        .launchIn(taskScope)
+
+    val tasks: Flow<List<TaskInfo>> = internalTaskData
+        .map { it.values }
+        .map { tasks ->
+            tasks.map {
+                TaskInfo(it, it.task.progress)
+            }
+        }
+
+    init {
+        Timber.tag(TAG).d("We have factories for %s", taskFactories.keys)
+    }
+
+    /**
+     * Don't re-use taskrequests, create new requests for each submission.
+     * They contain unique IDs.
+     */
+    fun submit(request: TaskRequest) {
+        if (!taskFactories.containsKey(request.type.java)) {
+            throw MissingTaskFactoryException(request::class)
+        }
+        Timber.tag(TAG).i("Task submitted: %s", request)
+        taskQueue.offer(request)
+    }
+
+    suspend fun cancel(requestId: UUID) = internalTaskData.updateSafely {
+        val taskState = values.singleOrNull {
+            it.request.id == requestId
+        } ?: throw IllegalArgumentException("No task found for request ID $requestId")
+
+        Timber.tag(TAG).w("Manually canceling %s", taskState)
+        taskState.deferred.cancel(cause = TaskCancelationException())
+    }
+
+    private suspend fun initTaskData(newRequest: TaskRequest) {
+        Timber.tag(TAG).d("Processing new task request: %s", newRequest)
+        val taskFactory = taskFactories[newRequest.type.java]
+        requireNotNull(taskFactory) { "No factory available for $newRequest" }
+
+        Timber.tag(TAG).v("Initiating task data for request: %s", newRequest)
+        val taskConfig = taskFactory.config
+        val task = taskFactory.taskProvider()
+
+        val deferred = taskScope.async(
+            start = CoroutineStart.LAZY
+        ) { task.run(newRequest.arguments) }
+
+        val activeTask = InternalTaskState(
+            request = newRequest,
+            createdAt = timeStamper.nowUTC,
+            config = taskConfig,
+            task = task,
+            deferred = deferred
+        )
+
+        internalTaskData.updateSafely {
+            val existingRequest = values.singleOrNull { it.request.id == newRequest.id }
+            if (existingRequest == null) {
+                Timber.tag(TAG).d("Added new pending task: %s", activeTask)
+                put(activeTask.id, activeTask)
+            } else {
+                Timber.tag(TAG).w(
+                    "TaskRequest was already used. Existing: %s\nNew request: %s",
+                    existingRequest, newRequest
+                )
+            }
+        }
+        taskScope.launch { processMap() }
+    }
+
+    private suspend fun processMap() = internalTaskData.updateSafely {
+        Timber.tag(TAG).d("Processing task data (count=%d)", size)
+        Timber.tag(TAG).v("Tasks before processing: %s", this.values)
+
+        // Procress all unprocessed finished tasks
+        values.toList()
```

Comment:
Could be moved to a dedicated function

Reply:
I don't like the idea of extracting methods that are called only from a single location. This method is not too complex and sufficiently documented in my opinion.

Reply:
I don't like the idea of extracting methods that are called only from a single location. This method is not too complex and sufficiently documented in my opinion.